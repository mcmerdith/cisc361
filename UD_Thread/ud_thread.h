/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);
