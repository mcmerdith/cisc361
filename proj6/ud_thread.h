/*
 * thread library function prototypes
 */

// Create a new thread from FUNCTION with id THREAD_ID and priority PRIORITY
void t_create(void (*function)(int), int thread_id, int priority);

// Terminate a thread. Semantically, this should be called at the end of every thread function
void t_terminate(void);

// Voluntarily release CPU control and be placed at the end of the thread queue
void t_yield(void);

// Initialize the thread library
void t_init(void);

// Shutdown and cleanup the thread library
void t_shutdown(void);

/**
 * semaphore
 */

// Create a new semaphore pointed to by sp with a count value of sem_count.
int sem_init(sem_t **sp, int sem_count);

// Current thread does a wait (P) on the specified semaphore.
void sem_wait(sem_t *sp);

// Current thread does a signal (V) on the specified semaphore.
// Follow the Mesa semantics (p. 9 of OSTEP Chapter 30 Condition Variables) where the thread that signals continues,
// and the first waiting (blocked) thread (if there is any) becomes ready.
void sem_signal(sem_t *sp);

// Free any memory related to the specified semaphore.
// If there are queued TCBs, they are moved to the ready queue, and this may imply that the semantics of the application might be incorrect.
void sem_destroy(sem_t **sp);

typedef void sem_t;