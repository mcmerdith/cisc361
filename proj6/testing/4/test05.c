/*
 * Test Program #5 - Send/Receive
 */

#include <stdio.h>
#include <string.h>
#include "ud_thread.h"

void sender(int thr_id)
{
    int i, j;
    char *test_message = "Test Message";

    printf("Message is: %s, length: %d\n", test_message, (int)strlen(test_message));

    for (i = j = 0; i < 3; i++, j++)
    {

        printf("[Pitch] - This is thread %d [%d]...\n", thr_id, j);
        send(2, test_message, strlen(test_message));
    }

    printf("Thread %d is done...\n", thr_id);
    t_terminate();
}

void catcher(int thr_id)
{
    int i, j;
    int snd_id;
    int len;
    char buffer[1024];

    snd_id = 1;
    for (i = j = 0; i < 3; i++, j++)
    {
        printf("[Catch] - This is thread %d [%d]...\n", thr_id, j);
        receive(&snd_id, buffer, &len);

        if (len)
        {
            printf("Catcher got [%s] from thread %d\n", buffer, snd_id);
        }
    }

    printf("Thread %d is done...\n", thr_id);
    t_terminate();
}

int main(void)
{

    int i;

    t_init();

    t_create(sender, 1, 1);
    t_create(catcher, 2, 1);
    t_yield();

    t_shutdown();

    return 0;
}
