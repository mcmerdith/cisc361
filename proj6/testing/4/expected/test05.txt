Message is: Test Message, length: 12
[Pitch] - This is thread 1 [0]...
[Pitch] - This is thread 1 [1]...
[Pitch] - This is thread 1 [2]...
Thread 1 is done...
[Catch] - This is thread 2 [0]...
Catcher got [Test Message] from thread 1
[Catch] - This is thread 2 [1]...
Catcher got [Test Message] from thread 1
[Catch] - This is thread 2 [2]...
Catcher got [Test Message] from thread 1
Thread 2 is done...
